dbuild/__init__.py

```diff
@@ -0,0 +1,247 @@
+
+
+def remove_container(docker_client, container, force=False):
+    """ Remove docker container """
+    return docker_client.remove_container(container=container, force=force)
+
+
+def create_docker_dir(dist, release):
+    """Create docker directory and populate it"""
+    PATH = os.path.dirname(os.path.abspath(__file__))
+    TMPL_ENV = Environment(
+        autoescape=False,
+        loader=FileSystemLoader(os.path.join(PATH, 'templates')),
+        trim_blocks=False)
+
+    # Create docker_dir - a temporary directory which will have Dockerfile and
+    # scripts to build the container.
+    docker_dir = mkdtemp()
+    dockerfile = os.path.join(docker_dir, 'Dockerfile')
+
+    ctxt = {'dist': dist, 'release': release,
+            'maintainer': 'dbuild, dbuild@test.com'}
+
+    # Write Dockerfile under docker_dir
+    with open(dockerfile, 'w') as d:
+        dockerdata = TMPL_ENV.get_template('dockerfile.jinja').render(ctxt)
+        d.write(dockerdata)
+
+    # Copy scripts under docker_dir
+    shutil.copytree(os.path.join(PATH, 'scripts'),
+                    os.path.join(docker_dir, 'scripts'))
+    return docker_dir
+
+def docker_build(build_dir, build_type, source_dir='source', force_rm=False,
+                 docker_url='unix://var/run/docker.sock', dist='ubuntu',
+                 release='trusty', extra_repos_file='repos',
+                 extra_repo_keys_file='keys', build_cache=True):
+    c = docker_client(docker_url)
+
+    docker_path = create_docker_dir(dist, release)
+
+    print "Starting %s Package Build" % build_type
+    image_tag = 'dbuild-' + dist + '/' + release
+    response = build_image(c, docker_path, tag=image_tag, nocache=not build_cache)
+    print response
+    command=['bash', '-c', '']
+
+    if os.path.exists(os.path.join(build_dir, extra_repos_file)):
+        command[2] += 'cp /build/%s \
+        /etc/apt/sources.list.d/dbuild-extra-repos.list && ' % extra_repos_file
+    else:
+        raise DbuildBuildFailedException('wrong extra repos file')
+
+    if os.path.exists(os.path.join(build_dir, extra_repo_keys_file)):
+        command[2] += 'apt-key add /build/%s && ' % extra_repo_keys_file
+    else:
+        raise DbuildBuildFailedException('wrong extra repo keys file')
+
+    command[2] += 'export DEBIAN_FRONTEND=noninteractive; apt-get -y update \
+                   && apt-get -y dist-upgrade && '
+
+    if build_type == 'source':
+        command[2] += 'dpkg-buildpackage -S -nc -uc -us'
+        cwd = '/build/' + source_dir
+    elif build_type == 'binary':
+        command[2] += "dpkg-source -x /build/*.dsc /build/pkgbuild/ && \
+                      cd /build/pkgbuild && \
+                      /usr/lib/pbuilder/pbuilder-satisfydepends && \
+                      dpkg-buildpackage"
+        cwd = '/build'
+    else:
+        shutil.rmtree(docker_path)
+        raise exceptions.DbuildBuildFailedException(
+            'Unknown build_type: %s' % build_type)
```

Comment:
This will fail to clean up the tmpdir.

Immediately after creating the tmpdir, start a try: block and put the cleanup in the finally: clause.